buildSrc/src/main/groovy/org/elasticsearch/gradle/VersionCollection.groovy

```diff
@@ -20,164 +20,336 @@
-        assert version != null : "getBWCSnapshotForPreviousMajor(): found no versions in the previous major"
-        return version
+    List<Version> getSnapshotVersionsWireCompatibleWithCurrent() {
+        List<Version> compatSnapshots = []
+        List<Version> allCompatVersions = getVersionsWireCompatibleWithCurrent()
+        if (allCompatVersions.contains(nextMinorSnapshot)) {
+            compatSnapshots.add(nextMinorSnapshot)
+        }
+        if (allCompatVersions.contains(stagedMinorSnapshot)) {
+            compatSnapshots.add(stagedMinorSnapshot)
+        }
+        if (allCompatVersions.contains(nextBugfixSnapshot)) {
+            compatSnapshots.add(nextBugfixSnapshot)
+        }
+        if (allCompatVersions.contains(maintenanceBugfixSnapshot)) {
+            compatSnapshots.add(maintenanceBugfixSnapshot)
+        }
+        // There was no wire compat for the 2.x line
+        compatSnapshots.removeAll {it.major == 2}
+
+        return compatSnapshots
     }
 
-    private Version getLastSnapshotWithMajor(int targetMajor) {
-        final String currentVersion = currentVersion.toString()
-        final int snapshotIndex = versions.findLastIndexOf {
-            it.major == targetMajor && it.before(currentVersion) && it.snapshot == buildSnapshot
+    /**
+     * Grabs the proper snapshot based on the name passed in. These names should correspond with gradle project names under bwc. If you
+     * are editing this if/else it is only because you added another project under :distribution:bwc. Do not modify this method or its
+     * reasoning for throwing the exception unless you are sure that it will not harm :distribution:bwc.
+     */
+    Version getCorrelatedNameToSnapshot(String snapshotProjectName) {
+        if (snapshotProjectName == 'next-minor-snapshot') {
+            return nextMinorSnapshot
+        } else if (snapshotProjectName == 'staged-minor-snapshot') {
+            return stagedMinorSnapshot
+        } else if (snapshotProjectName == 'maintenance-bugfix-snapshot') {
+            return maintenanceBugfixSnapshot
+        } else if (snapshotProjectName == 'next-bugfix-snapshot') {
+            return nextBugfixSnapshot
+        } else {
+            throw new InvalidUserDataException("Unsupported project name ${snapshotProjectName}")
         }
-        return snapshotIndex == -1 ? null : versions[snapshotIndex]
     }
 
-    private List<Version> versionsOnOrAfterExceptCurrent(Version minVersion) {
-        final String minVersionString = minVersion.toString()
-        return Collections.unmodifiableList(versions.findAll {
-            it.onOrAfter(minVersionString) && it != currentVersion
-        })
+    /**
+     * Uses basic logic about our releases to determine if this version has been previously released
+     */
+    private boolean isReleased(Version version) {
+        return version.revision > 0 || (version.revision > 1 && currentVersion.equals(Version.fromString("5.1.2")))
```

Comment:
Do we actually need this hack? None of the existing logic should be thinking any 5.x branch is current right? I understand it matters from the theoretical side, but in master 5.x version constants shouldn't even exist?

Reply:
++. I kept it in there just due to some initial notes I had taken about what released meant.